File: app/src/main/java/com/example/doshka/domain/model/Task.kt
package com.example.doshka.domain.model

import java.time.Instant

/**
 * Доменна модель задачі
 */
data class Task(
    val id: String,
    val title: String,
    val description: String? = null,
    val columnId: String,
    val position: Int,
    val priority: TaskPriority = TaskPriority.MEDIUM,
    val assignee: User? = null,
    val creator: User? = null,
    val deadline: Instant? = null,
    val estimatedHours: Float? = null,
    val actualHours: Float? = null,
    val tags: List<String> = emptyList(),
    val labels: List<Label> = emptyList(),
    val attachmentsCount: Int = 0,
    val commentsCount: Int = 0,
    val isDeleted: Boolean = false,
    val completedAt: Instant? = null,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
    val isSynced: Boolean = true // Статус синхронізації
) {
    /**
     * Перевіряє, чи задача прострочена
     */
    val isOverdue: Boolean
        get() = deadline != null && deadline.isBefore(Instant.now()) && completedAt == null

    /**
     * Перевіряє, чи наближається дедлайн (менше 24 годин)
     */
    val isDeadlineSoon: Boolean
        get() {
            if (deadline == null || completedAt != null) return false
            val dayFromNow = Instant.now().plusSeconds(24 * 60 * 60)
            return deadline.isBefore(dayFromNow) && !deadline.isBefore(Instant.now())
        }
}

/**
 * Пріоритет задачі
 */
enum class TaskPriority {
    CRITICAL,
    HIGH,
    MEDIUM,
    LOW;

    companion object {
        fun fromString(value: String): TaskPriority = when (value.lowercase()) {
            "critical" -> CRITICAL
            "high" -> HIGH
            "medium" -> MEDIUM
            "low" -> LOW
            else -> MEDIUM
        }
    }

    override fun toString(): String = name.lowercase()
}

File: server/app/schemas/task.py
"""
Схеми для задач та міток
"""
from datetime import datetime
from pydantic import BaseModel, Field
from .user import UserBrief


class LabelBase(BaseModel):
    """Базова схема мітки"""
    name: str = Field(min_length=1, max_length=100)
    color: str = Field(pattern="^#[0-9A-Fa-f]{6}$")  # HEX колір


class LabelCreate(LabelBase):
    """Схема створення мітки"""
    team_id: str


class LabelResponse(LabelBase):
    """Відповідь з даними мітки"""
    id: str
    team_id: str
    created_at: datetime

    class Config:
        from_attributes = True


class TaskBase(BaseModel):
    """Базова схема задачі"""
    title: str = Field(min_length=1, max_length=500)
    description: str | None = None
    priority: str = Field(default="medium", pattern="^(critical|high|medium|low)$")
    deadline: datetime | None = None
    estimated_hours: float | None = None
    tags: list[str] = []


class TaskCreate(TaskBase):
    """Схема створення задачі"""
    column_id: str
    assignee_id: str | None = None
    label_ids: list[str] = []


class TaskUpdate(BaseModel):
    """Схема оновлення задачі"""
    title: str | None = None
    description: str | None = None
    priority: str | None = None
    assignee_id: str | None = None
    deadline: datetime | None = None
    estimated_hours: float | None = None
    actual_hours: float | None = None
    tags: list[str] | None = None
    label_ids: list[str] | None = None
    version: int | None = None  # Для optimistic locking


class MoveTaskRequest(BaseModel):
    """Запит на переміщення задачі"""
    column_id: str
    position: int


class TaskResponse(TaskBase):
    """Відповідь з даними задачі"""
    id: str
    column_id: str
    position: int
    assignee_id: str | None = None
    assignee: UserBrief | None = None
    creator_id: str | None = None
    creator: UserBrief | None = None
    actual_hours: float | None = None
    labels: list[LabelResponse] = []
    attachments_count: int = 0
    comments_count: int = 0
    is_deleted: bool = False
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    version: int = 0  # Optimistic locking version

    class Config:
        from_attributes = True


class TaskBrief(BaseModel):
    """Коротка інформація про задачу"""
    id: str
    title: str
    priority: str
    deadline: datetime | None = None
    assignee: UserBrief | None = None

    class Config:
        from_attributes = True
